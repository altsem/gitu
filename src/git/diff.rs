use git2::Repository;
use itertools::Itertools;
use ratatui::{
    style::Style,
    text::{Line, Span, Text},
};
use similar::{udiff::UnifiedDiffHunk, Algorithm, ChangeTag, TextDiff};
use std::{fs, iter, ops::Range, path::PathBuf, rc::Rc, str};

use crate::{config::Config, Res};

#[derive(Debug, Clone)]
pub(crate) struct Diff {
    pub deltas: Vec<Delta>,
}

#[derive(Debug, Clone)]
pub(crate) struct Delta {
    pub file_header: String,
    pub old_file: PathBuf,
    pub new_file: PathBuf,
    pub hunks: Vec<Rc<Hunk>>,
    pub status: git2::Delta,
}

#[derive(Debug, Clone)]
pub(crate) struct Hunk {
    pub file_header: String,
    pub new_file: PathBuf,
    pub new_start: u32,
    pub header: String,
    pub content: Text<'static>,
}

#[derive(Debug)]
pub(crate) enum PatchMode {
    Normal,
    Reverse,
}

impl Hunk {
    pub(crate) fn format_patch(&self) -> String {
        format!("{}{}\n{}\n", &self.file_header, self.header, self.content)
    }

    pub(crate) fn format_line_patch(&self, line_range: Range<usize>, mode: PatchMode) -> String {
        let modified_content = self
            .content
            .lines
            .iter()
            .enumerate()
            .filter_map(|(i, line)| {
                let add = match mode {
                    PatchMode::Normal => '+',
                    PatchMode::Reverse => '-',
                };

                let remove = match mode {
                    PatchMode::Normal => '-',
                    PatchMode::Reverse => '+',
                };

                let patch_line = format!("{line}");

                if line_range.contains(&i) {
                    Some(patch_line)
                } else if patch_line.starts_with(add) {
                    None
                } else if let Some(stripped) = patch_line.strip_prefix(remove) {
                    Some(format!(" {}", stripped))
                } else {
                    Some(patch_line)
                }
            })
            .join("\n");

        format!(
            "{}{}\n{}\n",
            &self.file_header, self.header, modified_content
        )
    }

    pub(crate) fn first_diff_line(&self) -> u32 {
        self.content
            .lines
            .iter()
            .enumerate()
            .filter(|(_, line)| {
                let start = &line.spans.first().unwrap().content;
                start.starts_with('+') || start.starts_with('-')
            })
            .map(|(i, _)| i)
            .next()
            .unwrap_or(0) as u32
            + self.new_start
    }
}

pub(crate) fn convert_diff(
    config: &Config,
    repo: &Repository,
    diff: git2::Diff,
    workdir: bool,
) -> Res<Diff> {
    let mut deltas = vec![];

    diff.print(
        git2::DiffFormat::PatchHeader,
        |diffdelta, _maybe_hunk, line| {
            let line_content = str::from_utf8(line.content()).unwrap();
            let is_new_header = line_content.starts_with("diff")
                && line.origin_value() == git2::DiffLineType::FileHeader;

            if is_new_header {
                let mut delta = Delta {
                    file_header: line_content.to_string(),
                    old_file: path(&diffdelta.old_file()),
                    new_file: path(&diffdelta.new_file()),
                    hunks: vec![],
                    status: diffdelta.status(),
                };

                if let Ok(hunks) = diff_files(repo, diffdelta, workdir, config, &delta) {
                    delta.hunks = hunks;
                }

                deltas.push(delta);
            } else {
                let delta = deltas.last_mut().unwrap();
                delta.file_header.push_str(line_content);
            }

            true
        },
    )?;

    Ok(Diff { deltas })
}

fn diff_files(
    repo: &Repository,
    diffdelta: git2::DiffDelta<'_>,
    workdir: bool,
    config: &Config,
    delta: &Delta,
) -> Res<Vec<Rc<Hunk>>> {
    let old_content = read_blob(repo, &diffdelta.old_file())?;
    let new_content = if workdir {
        read_workdir(repo, &diffdelta.new_file())?
    } else {
        read_blob(repo, &diffdelta.new_file())?
    };

    diff_content(config, delta, old_content, new_content)
}

fn diff_content(
    config: &Config,
    delta: &Delta,
    old_content: String,
    new_content: String,
) -> Res<Vec<Rc<Hunk>>> {
    let text_diff = TextDiff::configure()
        .algorithm(Algorithm::Patience)
        .diff_lines(&old_content, &new_content);

    Ok(text_diff
        .unified_diff()
        .iter_hunks()
        .map(|hunk| {
            let formatted_hunk = format_hunk(config, &hunk, &text_diff);

            let new_start = hunk
                .header()
                .to_string()
                .strip_prefix("@@ -")
                .unwrap()
                .split(|c| c == ' ' || c == ',')
                .next()
                .unwrap()
                .parse()
                .unwrap();

            Rc::new(Hunk {
                file_header: delta.file_header.clone(),
                new_file: delta.new_file.clone(),
                new_start,
                header: format!("{}", hunk.header()),
                content: formatted_hunk,
            })
        })
        .collect::<Vec<_>>())
}

fn format_hunk<'diff, 'old, 'new, 'bufs>(
    config: &Config,
    hunk: &UnifiedDiffHunk<'diff, 'old, 'new, 'bufs, str>,
    text_diff: &'diff TextDiff<'old, 'new, 'bufs, str>,
) -> Text<'static>
where
    'diff: 'old + 'new,
{
    let formatted_hunk = hunk.ops().iter().flat_map(|op| {
        text_diff
            .iter_inline_changes(op)
            .map(|change| format_line_change(config, &change))
    });

    formatted_hunk.collect::<Vec<_>>().into()
}

fn format_line_change(config: &Config, change: &similar::InlineChange<str>) -> Line<'static> {
    let style = &config.style;

    let line_style = match change.tag() {
        ChangeTag::Equal => Style::new(),
        ChangeTag::Delete => (&style.line_removed).into(),
        ChangeTag::Insert => (&style.line_added).into(),
    };

    let some_emph = change.iter_strings_lossy().any(|(emph, _value)| emph);

    let spans = iter::once(Span::styled(format!("{}", change.tag()), line_style))
        .chain(change.iter_strings_lossy().map(|(emph, value)| {
            Span::styled(
                value.trim_end_matches('\n').to_string(),
                if some_emph {
                    if emph {
                        line_style.patch(&style.line_highlight.changed)
                    } else {
                        line_style.patch(&style.line_highlight.unchanged)
                    }
                } else {
                    line_style
                },
            )
        }))
        .collect::<Vec<_>>();

    Line::from(spans)
}

fn read_workdir(repo: &Repository, new_file: &git2::DiffFile<'_>) -> Res<String> {
    Ok(fs::read_to_string(
        repo.workdir()
            .expect("No workdir")
            .join(new_file.path().unwrap()),
    )?)
}

fn read_blob(repo: &Repository, file: &git2::DiffFile<'_>) -> Res<String> {
    let blob = repo.find_blob(file.id());
    blob.map(|blob| Ok(String::from_utf8(blob.content().to_vec())?))
        .unwrap_or(Ok("".to_string()))
}

fn path(file: &git2::DiffFile) -> PathBuf {
    file.path().unwrap().to_path_buf()
}
