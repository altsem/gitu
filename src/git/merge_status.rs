#[derive(Debug, Clone)]
pub(crate) struct MergeStatus {
    pub head: String,
}
