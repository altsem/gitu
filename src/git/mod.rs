use git2::Repository;
use itertools::Itertools;

use self::{commit::Commit, diff::Diff, merge_status::MergeStatus, rebase_status::RebaseStatus};
use crate::{config::Config, git2_opts, Res};
use std::{
    ffi::OsStr,
    fs,
    path::Path,
    process::Command,
    str::{self},
};

pub(crate) mod commit;
pub(crate) mod diff;
pub(crate) mod merge_status;
pub(crate) mod rebase_status;

// TODO Use only plumbing commands

pub(crate) fn rebase_status(repo: &Repository) -> Res<Option<RebaseStatus>> {
    let dir = repo.workdir().expect("No workdir");
    let mut rebase_onto_file = dir.to_path_buf();
    rebase_onto_file.push(".git/rebase-merge/onto");

    let mut rebase_head_name_file = dir.to_path_buf();
    rebase_head_name_file.push(".git/rebase-merge/head-name");

    match fs::read_to_string(&rebase_onto_file) {
        Ok(content) => {
            let onto_hash = content.trim().to_string();
            Ok(Some(RebaseStatus {
                onto: branch_name(dir, &onto_hash)?.unwrap_or_else(|| onto_hash[..7].to_string()),
                head_name: fs::read_to_string(rebase_head_name_file)?
                    .trim()
                    .strip_prefix("refs/heads/")
                    .unwrap()
                    .to_string(),
                // TODO include log of 'done' items
            }))
        }
        Err(err) => {
            log::warn!(
                "Couldn't read {}, due to {}",
                rebase_onto_file.to_string_lossy(),
                err
            );
            Ok(None)
        }
    }
}

pub(crate) fn merge_status(repo: &Repository) -> Res<Option<MergeStatus>> {
    let dir = repo.workdir().expect("No workdir");
    let mut merge_head_file = dir.to_path_buf();
    merge_head_file.push(".git/MERGE_HEAD");

    match fs::read_to_string(&merge_head_file) {
        Ok(content) => {
            let head = content.trim().to_string();
            Ok(Some(MergeStatus {
                head: branch_name(dir, &head)?.unwrap_or(head[..7].to_string()),
            }))
        }
        Err(err) => {
            log::warn!(
                "Couldn't read {}, due to {}",
                merge_head_file.to_string_lossy(),
                err
            );
            Ok(None)
        }
    }
}

// TODO replace with libgit2
fn branch_name(dir: &Path, hash: &str) -> Res<Option<String>> {
    let out = Command::new("git")
        .args(["for-each-ref", "--format", "%(objectname) %(refname:short)"])
        .current_dir(dir)
        .output()?
        .stdout;

    Ok(str::from_utf8(&out)?
        .lines()
        .find(|line| line.starts_with(hash))
        .map(|line| line.split(' ').nth(1).unwrap().to_string()))
}

pub(crate) fn diff_unstaged(config: &Config, repo: &Repository) -> Res<Diff> {
    let diff = repo.diff_index_to_workdir(None, Some(&mut git2_opts::diff(repo)?))?;
    diff::convert_diff(config, repo, diff, true)
}

pub(crate) fn diff_staged(config: &Config, repo: &Repository) -> Res<Diff> {
    let opts = &mut git2_opts::diff(repo)?;

    let diff = match repo.head() {
        Ok(head) => repo.diff_tree_to_index(Some(&head.peel_to_tree()?), None, Some(opts))?,
        Err(_) => repo.diff_tree_to_index(None, None, Some(opts))?,
    };

    diff::convert_diff(config, repo, diff, false)
}

pub(crate) fn show(config: &Config, repo: &Repository, reference: &str) -> Res<Diff> {
    let object = &repo.revparse_single(reference)?;

    let commit = object.peel_to_commit()?;
    let tree = commit.tree()?;
    let parent_tree = commit
        .parents()
        .next()
        .and_then(|parent| parent.tree().ok());

    let diff = repo.diff_tree_to_tree(
        parent_tree.as_ref(),
        Some(&tree),
        Some(&mut git2_opts::diff(repo)?),
    )?;

    diff::convert_diff(config, repo, diff, false)
}

pub(crate) fn show_summary(repo: &Repository, reference: &str) -> Res<Commit> {
    let object = &repo.revparse_single(reference)?;
    let commit = object.peel_to_commit()?;

    let author = commit.author();
    let name = author.name().unwrap_or("");
    let email = commit
        .author()
        .email()
        .map(|email| format!("<{}>", email))
        .unwrap_or("".to_string());

    let message = commit
        .message()
        .unwrap_or("")
        .to_string()
        .lines()
        .map(|line| format!("    {}", line))
        .join("\n");

    let offset = chrono::FixedOffset::east_opt(author.when().offset_minutes() * 60).unwrap();
    let time = chrono::DateTime::with_timezone(
        &chrono::DateTime::from_timestamp(author.when().seconds(), 0).unwrap(),
        &offset,
    );

    let details = format!(
        "Author: {}\nDate:   {}\n\n{}",
        [name, &email].join(" "),
        time.to_rfc2822(),
        message
    );

    Ok(Commit {
        hash: commit.id().to_string(),
        details,
    })
}

pub(crate) fn stage_file_cmd(file: &OsStr) -> Command {
    git([OsStr::new("add"), file])
}
pub(crate) fn stage_patch_cmd() -> Command {
    git(["apply", "--cached"])
}
pub(crate) fn stage_line_cmd() -> Command {
    git(["apply", "--cached", "--recount"])
}
pub(crate) fn unstage_file_cmd(file: &OsStr) -> Command {
    git([OsStr::new("restore"), OsStr::new("--staged"), file])
}
pub(crate) fn unstage_patch_cmd() -> Command {
    git(["apply", "--cached", "--reverse"])
}
pub(crate) fn unstage_line_cmd() -> Command {
    git(["apply", "--cached", "--reverse", "--recount"])
}
pub(crate) fn discard_unstaged_patch_cmd() -> Command {
    git(["apply", "--reverse"])
}
pub(crate) fn discard_branch(branch: &OsStr) -> Command {
    git([OsStr::new("branch"), OsStr::new("-d"), branch])
}
pub(crate) fn commit_fixup_cmd(reference: &OsStr) -> Command {
    git([OsStr::new("commit"), OsStr::new("--fixup"), reference])
}
pub(crate) fn reset_soft_cmd(reference: &OsStr) -> Command {
    git([OsStr::new("reset"), OsStr::new("--soft"), reference])
}
pub(crate) fn reset_mixed_cmd(reference: &OsStr) -> Command {
    git([OsStr::new("reset"), OsStr::new("--mixed"), reference])
}
pub(crate) fn reset_hard_cmd(reference: &OsStr) -> Command {
    git([OsStr::new("reset"), OsStr::new("--hard"), reference])
}
pub(crate) fn checkout_file_cmd(file: &OsStr) -> Command {
    git([OsStr::new("checkout"), OsStr::new("--"), file])
}

pub(crate) fn git<I, S>(args: I) -> Command
where
    I: IntoIterator<Item = S>,
    S: AsRef<OsStr>,
{
    let mut cmd = Command::new("git");
    cmd.args(args);
    cmd
}
