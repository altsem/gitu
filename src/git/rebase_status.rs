#[derive(Debug, Clone)]
pub(crate) struct RebaseStatus {
    pub onto: String,
    pub head_name: String,
}
