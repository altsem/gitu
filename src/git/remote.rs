use std::{ops::Deref, str};

use git2::{Branch, Remote, Repository};

use crate::{git, Res};

use super::{Error, Utf8Error};

pub(crate) fn get_upstream(repo: &Repository) -> Res<Option<Branch>> {
    match git::get_current_branch(repo)?.upstream() {
        Ok(v) => Ok(Some(v)),
        Err(e) if e.class() == git2::ErrorClass::Config => Ok(None),
        Err(e) => Err(Error::GetCurrentBranchUpstream(e)),
    }
}

/// If the branch has an upstream, returns the remote name and branch name in that order.
/// Returns "." as remote if the current branch has no remote upstream.
///
/// Branch references would be used like this (in Magit)
///
/// // Remote branch
/// git … push -v origin feature-branch\:refs/heads/feature-branch
/// git … pull origin refs/heads/feature-branch
/// git … rebase --autostash origin/feature-branch
///
/// // Local branch
/// git … push -v . feature-branch\:refs/heads/main
/// git … pull . refs/heads/main
/// git … rebase --autostash main
pub(crate) fn get_upstream_components(repo: &Repository) -> Res<Option<(String, String)>> {
    let Some(upstream) = get_upstream(repo)? else {
        return Ok(None);
    };

    let branch = String::from_utf8(upstream.get().shorthand_bytes().to_vec())
        .map_err(Utf8Error::String)
        .map_err(Error::BranchNameUtf8)?;

    if upstream.get().is_remote() {
        let branch_full = str::from_utf8(upstream.get().name_bytes())
            .map_err(Utf8Error::Str)
            .map_err(Error::BranchNameUtf8)?;
        let remote = String::from_utf8(
            repo.branch_remote_name(branch_full)
                .map_err(Error::GetRemote)?
                .deref()
                .to_vec(),
        )
        .map_err(Utf8Error::String)
        .map_err(Error::RemoteNameUtf8)?;

        let remote_prefix = format!("{}/", remote);
        Ok(Some((remote, branch.replace(&remote_prefix, ""))))
    } else {
        Ok(Some((".".into(), branch)))
    }
}

pub(crate) fn get_upstream_shortname(repo: &Repository) -> Res<Option<String>> {
    let Some(upstream) = get_upstream(repo)? else {
        return Ok(None);
    };
    Ok(Some(
        String::from_utf8(upstream.get().shorthand_bytes().to_vec())
            .map_err(Utf8Error::String)
            .map_err(Error::GetCurrentBranchUpstreamUtf8)?,
    ))
}

pub(crate) fn get_push_remote(repo: &Repository) -> Res<Option<String>> {
    let push_remote_cfg = head_push_remote_cfg(repo)?;
    let config = repo.config().map_err(Error::ReadGitConfig)?;
    let result = match config.get_entry(&push_remote_cfg) {
        Ok(entry) => Ok(Some(
            String::from_utf8(entry.value_bytes().to_vec())
                .map_err(Utf8Error::String)
                .map_err(Error::ReadGitConfigUtf8)?,
        )),
        Err(e) if e.class() == git2::ErrorClass::Config => get_default_push_remote(repo),
        Err(e) => Err(Error::ReadGitConfig(e)),
    };

    result
}

pub(crate) fn get_default_push_remote(repo: &Repository) -> Res<Option<String>> {
    let push_default_cfg = "remote.pushDefault";
    let config = repo.config().map_err(Error::ReadGitConfig)?;
    let result = match config.get_entry(push_default_cfg) {
        Ok(entry) => Ok(Some(
            String::from_utf8(entry.value_bytes().to_vec())
                .map_err(Utf8Error::String)
                .map_err(Error::ReadGitConfigUtf8)?,
        )),
        Err(e) if e.class() == git2::ErrorClass::Config => Ok(None),
        Err(e) => Err(Error::ReadGitConfig(e)),
    };

    result
}

pub(crate) fn set_push_remote(repo: &Repository, remote: Option<&Remote>) -> Res<()> {
    let push_remote_cfg = head_push_remote_cfg(repo)?;
    let mut config = repo.config().map_err(Error::ReadGitConfig)?;
    match remote {
        None => {
            config
                .remove(&push_remote_cfg)
                .map_err(Error::DeleteGitConfig)?;
        }
        Some(remote) => {
            config
                .set_str(
                    &push_remote_cfg,
                    str::from_utf8(remote.name_bytes().ok_or(Error::RemoteHasNoName)?)
                        .map_err(Utf8Error::Str)
                        .map_err(Error::RemoteNameUtf8)?,
                )
                .map_err(Error::SetGitConfig)?;
        }
    }
    Ok(())
}

pub(crate) fn head_push_remote_cfg(repo: &Repository) -> Res<String> {
    let branch = git::get_current_branch_name(repo)?;
    let push_remote_cfg = format!("branch.{branch}.pushRemote");
    Ok(push_remote_cfg)
}
