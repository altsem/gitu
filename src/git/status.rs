#[derive(Debug, PartialEq, Eq)]
pub(crate) struct Status {
    pub branch_status: BranchStatus,
    pub files: Vec<StatusFile>,
}

#[derive(Debug, PartialEq, Eq)]
pub(crate) struct BranchStatus {
    pub local: Option<String>,
    pub remote: Option<String>,
    pub ahead: u32,
    pub behind: u32,
}

#[derive(Debug, PartialEq, Eq)]
pub(crate) struct StatusFile {
    pub status_code: [char; 2],
    pub path: String,
    pub new_path: Option<String>,
}

impl StatusFile {
    pub fn is_unmerged(&self) -> bool {
        matches!(
            self.status_code,
            ['D', 'D']
                | ['A', 'U']
                | ['U', 'D']
                | ['U', 'A']
                | ['D', 'U']
                | ['A', 'A']
                | ['U', 'U']
        )
    }

    pub fn is_untracked(&self) -> bool {
        self.status_code == ['?', '?']
    }
}
