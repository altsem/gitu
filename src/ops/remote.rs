use std::{process::Command, rc::Rc};

use crate::{
    items::TargetData,
    state::{PromptParams, State},
    term::Term,
    Res,
};

use super::{Action, OpTrait};

pub(crate) struct AddRemote;
impl OpTrait for AddRemote {
    fn get_action(&self, _target: Option<&TargetData>) -> Option<Action> {
        Some(Rc::new(
            |state: &mut crate::state::State, term: &mut Term| {
                let remote_name = state.prompt(
                    term,
                    &PromptParams {
                        prompt: "Remote name",
                        ..Default::default()
                    },
                )?;

                let remote_url = state.prompt(
                    term,
                    &PromptParams {
                        prompt: "Remote url",
                        ..Default::default()
                    },
                )?;

                add_remote_with_name(state, term, &remote_name, &remote_url)?;

                Ok(())
            },
        ))
    }

    fn display(&self, _state: &State) -> String {
        "add remote".to_string()
    }
}

pub(crate) struct RenameRemote;
impl OpTrait for RenameRemote {
    fn get_action(&self, _target: Option<&TargetData>) -> Option<Action> {
        Some(Rc::new(
            |state: &mut crate::state::State, term: &mut Term| {
                let remote_name = state.prompt(
                    term,
                    &PromptParams {
                        prompt: "Rename remote",
                        ..Default::default()
                    },
                )?;

                let new_remote_name = state.prompt(
                    term,
                    &PromptParams {
                        prompt: "Rename to",
                        ..Default::default()
                    },
                )?;

                rename_remote(state, term, &remote_name, &new_remote_name)?;
                Ok(())
            },
        ))
    }

    fn display(&self, _state: &State) -> String {
        "rename remote".to_string()
    }
}

fn rename_remote(
    state: &mut State,
    term: &mut ratatui::Terminal<crate::term::TermBackend>,
    remote_name: &str,
    new_remote_name: &str,
) -> Res<()> {
    let mut cmd = Command::new("git");
    cmd.args(["remote", "rename", remote_name, new_remote_name]);

    state.close_menu();
    state.run_cmd(term, &[], cmd)?;
    Ok(())
}

fn add_remote_with_name(
    state: &mut State,
    term: &mut Term,
    remote_name: &str,
    remote_url: &str,
) -> Res<()> {
    let mut cmd = Command::new("git");
    cmd.args(["remote", "add", remote_name, remote_url]);

    state.close_menu();
    state.run_cmd(term, &[], cmd)?;
    Ok(())
}
