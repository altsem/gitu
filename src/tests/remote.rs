use git2::{Buf, Error, Repository};

use crate::git::remote::*;
use crate::tests::helpers::{run, RepoTestContext};

fn get_head_name(repo: &Repository) -> String {
    repo.head().unwrap().name().unwrap().into()
}

fn get_branch_remote(repo: &Repository) -> Result<Buf, Error> {
    repo.branch_upstream_remote(&get_head_name(repo))
}

fn get_branch_remote_str(repo: &Repository) -> String {
    get_branch_remote(repo).unwrap().as_str().unwrap().into()
}

#[test]
fn set_push_remote_basic() {
    let ctx = RepoTestContext::setup_clone();

    let repo = ctx.local_repo;

    let push_remote = get_push_remote(&repo).unwrap();
    assert_eq!(push_remote, None);

    let remote_name = get_branch_remote_str(&repo);
    let remote = repo.find_remote(&remote_name).unwrap();
    set_push_remote(&repo, Some(&remote)).unwrap();
    let push_remote = get_push_remote(&repo).unwrap();
    assert_eq!(push_remote, Some(remote_name));

    set_push_remote(&repo, None).unwrap();
    let push_remote = get_push_remote(&repo).unwrap();
    assert_eq!(push_remote, None);
}

#[test]
fn get_upstream_basic() {
    let ctx = RepoTestContext::setup_clone();
    let repo = ctx.local_repo;
    let upstream = get_upstream(&repo).unwrap().unwrap();
    assert_eq!(upstream.name().unwrap().unwrap(), "origin/main");

    run(ctx.dir.path(), &["git", "branch", "--unset-upstream"]);
    let upstream = get_upstream(&repo).unwrap();
    assert!(upstream.is_none());
}

#[test]
fn get_upstream_components_of_remote_branch() {
    let ctx = RepoTestContext::setup_clone();
    let repo = ctx.local_repo;

    let (remote, branch) = get_upstream_components(&repo).unwrap().unwrap();
    assert_eq!(remote, "origin");
    assert_eq!(branch, "main");
}

#[test]
fn get_upstream_components_of_feature_branch() {
    let ctx = RepoTestContext::setup_clone();
    let repo = ctx.local_repo;
    run(ctx.dir.path(), &["git", "checkout", "-b", "feature-branch"]);
    run(
        ctx.dir.path(),
        &["git", "branch", "--set-upstream-to", "main"],
    );

    let (remote, branch) = get_upstream_components(&repo).unwrap().unwrap();
    assert_eq!(remote, ".");
    assert_eq!(branch, "main");
}
