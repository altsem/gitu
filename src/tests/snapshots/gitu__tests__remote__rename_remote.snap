---
source: src/tests/remote.rs
expression: ctx.redact_buffer()
---
▌On branch main                                                                 |
▌Your branch is up to date with 'origin2/main'.                                 |
                                                                                |
 Recent commits                                                                 |
 b66a0bf main origin2/main add initial-file                                     |
                                                                                |
                                                                                |
                                                                                |
                                                                                |
                                                                                |
                                                                                |
                                                                                |
                                                                                |
                                                                                |
                                                                                |
                                                                                |
                                                                                |
                                                                                |
────────────────────────────────────────────────────────────────────────────────|
$ git remote rename origin origin2                                              |
styles_hash: b8f048b49ddbac3f
